### PR TITLE
[JUJU-1054] Ensure to convert pod status to juju status consistent across operato…

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR ensures
1. convert pod status to juju status consistently across operator and sidecar applications;
2. decide sidecar application status based on the `statefulset/daemonset/deployment`(daemonset/deployment are not implemenmented because sidecar charms do not support these two workload types) status rather than the service status;

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

## QA steps

```console
# deploy sidecar app with a wrong image path
$ juju deploy snappass-test --resource snappass-image=snappass-test1 --resource redis-image=redis

$ juju status --relations --color --storage -m k1:t1
Model  Controller  Cloud/Region        Version  SLA          Timestamp
t1     k1          microk8s/localhost  2.9.30   unsupported  17:34:18+10:00

App            Version  Status   Scale  Charm          Channel  Rev  Address         Exposed  Message
snappass-test           waiting      1  snappass-test  stable     9  10.152.183.168  no       installing agent

Unit              Workload  Agent  Address       Ports  Message
snappass-test/0*  error     idle   10.1.245.129         unknown container reason "ImagePullBackOff": Back-off pulling image "snappass-test1"

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1946574

### CODE REVIEW

File: caas/kubernetes/provider/resources/events.go [hunk 1]
```diff
@@ -15,20 +15,31 @@ import (
 
 const maxEventsToPage = 100
 
+func listEventsForObject(
```
Comment (wallyworld) [inline caas/kubernetes/provider/resources/events.go:18]: We should just export this one and delete the one below. Having 2 methods with the same name is confusing.
From what I can see the place where we call the one below
```
return resources.ListEventsForObject(context.TODO(), objName, objKind, k.client().CoreV1().Events(k.namespace))
```
can be rewritten as
```
return resources.ListEventsForObject(context.TODO(), k.client(), k.namespace, objName, objKind)
```

File: caas/kubernetes/provider/resources/pod.go [hunk 2]
```diff
@@ -120,46 +122,198 @@ func (p *Pod) Delete(ctx context.Context, client kubernetes.Interface) error {
...
+		PodReasonContainersNotReady:       status.Maintenance,
+		PodReasonContainersNotInitialized: status.Maintenance,
+	}
+
+	podScheduledReasonsMap = map[string]status.Status{
+		corev1.PodReasonUnschedulable: status.Blocked,
+	}
+)
+
+// PodToJujuStatus takes a Kubernetes pod and translate's it to a known Juju
```
Comment (wallyworld) [inline caas/kubernetes/provider/resources/pod.go:165]: translates

File: caas/kubernetes/provider/resources/pod.go [hunk 3]
```diff
@@ -120,46 +122,198 @@ func (p *Pod) Delete(ctx context.Context, client kubernetes.Interface) error {
...
-			statusMessage = eventList[count-1].Message
+			defaultStatusMessage = eventList[count-1].Message
 		}
 	}
-	return statusMessage, jujuStatus, since, nil
+	return defaultStatusMessage, status.Unknown, since, nil
+}
+
+// interrogatePodContainerStatus combs a set of container statuses. If a
+// container is found to be in an error state it's error message and true are
```
Comment (wallyworld) [inline caas/kubernetes/provider/resources/pod.go:271]: its

File: caas/kubernetes/provider/application/application.go [hunk 4]
```diff
@@ -1089,6 +1090,40 @@ func (a *app) Service() (*caas.Service, error) {
...
+		if jujuStatus == status.Waiting {
+			events, err := ss.Events(ctx, client)
+			if err != nil {
+				return "", jujuStatus, now, errors.Trace(err)
+			}
+			// Take the most recent event.
+			if count := len(events); count > 0 {
+				evt := events[count-1]
+				if evt.Type == corev1.EventTypeWarning && evt.Reason == "FailedCreate" {
+					jujuStatus = status.Blocked
```
Comment (wallyworld) [inline caas/kubernetes/provider/application/application.go:1116]: error is what we'd want here.
blocked is usually reserved as being for the charm to set

File: caas/kubernetes/provider/application/application.go [hunk 5]
```diff
@@ -1089,6 +1090,40 @@ func (a *app) Service() (*caas.Service, error) {
 	}, nil
 }
 
+func (a *app) computeStatus(ctx context.Context, client kubernetes.Interface, now time.Time) (string, status.Status, time.Time, error) {
+	jujuStatus := status.Waiting
```
Comment (wallyworld) [inline caas/kubernetes/provider/application/application.go:1094]: allocating is what we'd want here.
waiting is usually reserved as being for the charm to set

File: caas/kubernetes/provider/application/application.go [hunk 6]
```diff
@@ -1089,6 +1090,40 @@ func (a *app) Service() (*caas.Service, error) {
...
+	switch a.deploymentType {
+	case caas.DeploymentStateful:
+		ss, err := a.getStatefulSet()
+		if err != nil {
+			return "", jujuStatus, now, errors.Trace(err)
+		}
+		if ss.GetDeletionTimestamp() != nil {
+			jujuStatus = status.Terminated
+		} else if ss.Spec.Replicas != nil && ss.Status.ReadyReplicas == *ss.Spec.Replicas {
+			jujuStatus = status.Active
```
Comment (wallyworld) [inline caas/kubernetes/provider/application/application.go:1104]: an application can be active even if all pods are not yet ready
i think we'd want "allocating" if ReadyReplicas is 0 and "active" if there's at least 1
if something different is needed, then it's up to the charm to set application status itself

**Review by wallyworld — APPROVED** [review]
A couple of small things to look at. The application status doesn't quite match what we'd want juju to do.

File: caas/kubernetes/provider/application/application.go (hunk 4 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/application/application.go:1116]: This logic was copied from operator k8s code. 
yeah, agreed, it should be Error.

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**